File: patterns/2020_09_13/source/NormalImpl.py
from typing import Any, Iterable
import sys


class MyClass:

    def draw(self, os, position) -> None:
        print(f"{' '*position}MyClass", file=os)


def draw(document: Iterable[Any], os=sys.stdout, position=0):
    print(f"{' '*position}<document>", file=os)
    for e in document:
        e.draw(os, position+2)
    print(f"{' '*position}</document>", file=os)


if __name__ == "__main__":
    document = [
        MyClass()
    ]
    draw(document)

File: patterns/2020_09_13/source/NormalImpl.cpp
#include <iostream>
#include <vector>
#include <memory>

class object_t
{
public:
    virtual ~object_t() = default;
    virtual void draw(std::ostream &, std::size_t) const = 0;
};

using document_t = std::vector<std::shared_ptr<object_t>>;

void draw(document_t const &doc, std::ostream &os, std::size_t position)
{
    os << std::string(position, ' ') << "<document>" << std::endl;
    for (const auto &e : doc)
    {
        e->draw(os, position + 2);
    }
    os << std::string(position, ' ') << "</document>" << std::endl;
}

class my_class_t : public object_t
{
public:
    my_class_t() = default;
    void draw(std::ostream &os, std::size_t position) const override
    {
        os << std::string(position, ' ') << "my_class_t" << std::endl;
    }
};

int main(int argc, char **argv)
{
    document_t document;
    document.emplace_back(std::make_shared<my_class_t>());
    draw(document, std::cout, 0);
    return 0;
}
